### PR TITLE
_WD_Navigate - support for local file

### PR DESCRIPTION
**Pull request by mlipok**

### CODE REVIEW

File: wd_core.au3 [hunk 1]
```diff
@@ -364,13 +364,13 @@ EndFunc   ;==>_WD_Timeouts
 ; Description ...: Navigate to the designated URL.
 ; Syntax ........: _WD_Navigate($sSession, $sURL)
 ; Parameters ....: $sSession - Session ID from _WD_CreateSession
-;                  $sURL     - Destination URL
+;                  $sURL     - - Destination URL or local file full path to open in browser
```
[the remark this answers is not in the document]
Comment (mlipok) [reply wd_core.au3:367]: Do it please

File: wd_core.au3 [hunk 2]
```diff
@@ -380,6 +380,13 @@ Func _WD_Navigate($sSession, $sURL)
 	Local Const $sFuncName = "_WD_Navigate"
 	Local $sResponse = __WD_Post($_WD_BASE_URL & ":" & $_WD_PORT & "/session/" & $sSession & "/url", '{"url":"' & $sURL & '"}')
 
+	If StringInStr($sURL,'\') Then
+		If Not FileExists($sURL) Then
+			Return SetError(__WD_Error($sFuncName, $_WD_ERROR_InvalidValue, "File not exist: " & $sURL), 0, 0)
+		EndIf
+		$sURL= "file:///" & StringReplace($sURL, "\", "/")
+	EndIf
+
```
[the remark this answers is not in the document]
Comment (mlipok) [reply wd_core.au3:390]: Sure my fault.
I'm laying to bed, and writing frome phone.
 please change it and merge if it will be fine.

**Review by mlipok — COMMENTED** [review]
Going to bed leaving comments on my smartphone.

**Comment by mlipok** [comment]
No.
At least not in this case.
I can elaborate later.

File: wd_core.au3 (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlipok) [reply wd_core.au3]: done

File: wd_core.au3 (hunk 2 above)
[the remark this answers is not in the document]
Comment (mlipok) [reply wd_core.au3]: done

**Comment by mlipok** [comment]
> No. At least not in this case. I can elaborate later.

**Reason 1.**
As so far there was no other function to open files from disc.

**Reason 2.**
_WD_Navigate() should support opening files natively in an easy way so take a look for the next resosns

**Reason 3.**
User should not bother about proper file formating I mean do not need to add prefix `"file:///" &` each time and do not need to change `StringReplace($sURL, "\", "/")`

You can as why ? ..... Because if you copy local html file full path to clipboard and past it to any browser then browser do not bother you about proper usage just browser check it, and fix.   The same way WD UDF should act.

**Reason 4.**
as to checking if file exist 
web browsers do the same

and my famous motto:
 "Any program that accepts data from a user must include code to validate that data before sending it to the data store. You cannot rely on the data store, ...., or even your programming language to notify you of problems. You must check every byte entered by your users, making sure that data is the correct type for its field and that required fields are not empty."

https://docs.microsoft.com/en-us/sql/ado/guide/data/anticipating-errors?redirectedfrom=MSDN&view=sql-server-ver15

**Comment by Danp2** [comment]
> as to checking if file exist web browsers do the same

Exactly my point. I believe that _WD_Navigate should not be performing this check. Leave it up to the browser to provide the feedback. 

Try it. What is the result if you allow the browser to handle it? What type of feedback is returned by the webdriver?

>  As so far there was no other function to open files from disc.

True, but that could easily be rectified by moving this code to a separate function. But let's hold off on that until we resolve the rest of this conversation,

> and my favorite motto:

This is great. But I would argue that it doesn't necessarily apply here. Why? Because we aren't dealing with end users. This is a library of functions designed for script writers to use. And it is their responsibility to validate all incoming data, not the functions they utilize from a 3rd party UDF.

**Comment by mlipok** [comment]
> This is great. But I would argue that it doesn't necessarily apply here. Why? Because we aren't dealing with end users. This is a library of functions designed for script writers to use. And it is their responsibility to validate all incoming data, not the functions they utilize from a 3rd party UDF.

In my approach, each UDF should take care of what parameters are passed to the function, and it does not matter whether the UDF concerns data management in the database, but any processing, because the lack of control at every minor step will result in the same error - writing incorrect result data to any target data container (SQL, INI, TXT Log, XLS report, PDF).

I hope to make you follow the same approach in this UDF as well.

**Comment by Danp2** [comment]
> I do not want to create separate function, this was only notice why I propose to add new features to _WD_Navigate()

Why? Can you not foresee the need for this anywhere else?

> So it is UDF role to check and parse proper windows FileFullPath 

1) In my testing, only the `StringReplace($sFilePath, "\", "/")` is required to make this work. 

2) From my limited research, it appears that a backslash is a valid character in a URL

3) I'll repeat my belief that FileExists() check doesn't belong here

> I hope to make you follow the same approach in this UDF as well.

Good luck 😛

**Comment by mlipok** [comment]
> Why? Can you not foresee the need for this anywhere else?

It seems natural for me to be in this place.
even: https://www.w3.org/TR/webdriver/#navigate-to
Mention about using navigate to file `8. If url is special except for file and current URL and URL do not have the same absolute URL`

This mean that webdriver support navigation to file in Navigate functions .
Thus I stand that _WD_Navigate() should support navigation to file.

**Comment by mlipok** [comment]
> * In my testing, only the `StringReplace($sFilePath, "\", "/")` is required to make this work.
> * From my limited research, it appears that a backslash is a valid character in a URL

so is valid.... but required

> * I'll repeat my belief that FileExists() check doesn't belong here

in such case:
```
Func _WD_Navigate($sSession, $sURL)
	Local Const $sFuncName = "_WD_Navigate"

	If StringInStr($sURL, '\') Then
;~ 		If Not FileExists($sURL) Then
;~ 			Return SetError(__WD_Error($sFuncName, $_WD_ERROR_NotFound, "File not exist: " & $sURL), 0, 0)
;~ 		EndIf
		$sURL = "file:///" & StringReplace($sURL, "\", "/")
	EndIf
	.......
	.......
```


this example snippet:
```
	Local $sFilePath = @ScriptDir & "\NotExistenFile.html"
	_WD_Navigate($sSession, $sFilePath)
	If @error Then Then Return

```
did not Return because @error is not set regadless of the fact that in browser I see: ERR_FILE_NOT_FOUND

**Comment by mlipok** [comment]
> Good luck

I am patient and I think with time I think I will succeed

**Comment by mlipok** [comment]
> did not Return because @error is not set regadless of the fact that in browser I see: ERR_FILE_NOT_FOUND

checked with this small mod:
```
Func __WD_Post($sURL, $sData)
		.......
		.......
		.......
			If $iErr Then
				$iResult = $_WD_ERROR_SendRecv
				$sResponseText = $WD_WinHTTPTimeoutMsg
			Else
				ConsoleWrite("! $iErr = " & @CRLF&$iErr& @CRLF)
				ConsoleWrite("! $sResponseText = " & @CRLF&$sResponseText& @CRLF)

				MsgBox($MB_OK + $MB_TOPMOST + $MB_ICONINFORMATION, "Information #" & @ScriptLineNumber, $iErr )
				__WD_DetectError($iErr, $sResponseText)
				$iResult = $iErr
			EndIf
		.......
		.......

```
I get such results:
```
__WD_Post: URL=HTTP://127.0.0.1:9515/session/da3434f4a9cc00f2924074bb8905cdc3/url; $sData={"url":"file:///Z:/!!!_SVN_AU3/UDF_Forum/Other_Members/Test/WebDriver/NotExistenFile.html"}
! $iErr = 
0
! $sResponseText = 
{"value":null}

```
It looks like WebDriver at least chromedriver.exe does not give any error in this case.
Thus IMHO it is WD UDF role to check if the file exist.

**Comment by Danp2** [comment]
>  Thus I stand that _WD_Navigate() should support navigation to file.

I never said or even implied that it shouldn't. We just disagree on your proposed implementation. 

- I believe that we both agree that the existing function works as long as $sURL is formatted correctly by the calling routine. 
- You want to add FileExists() checks; I don't believe that belongs in this function.
- You want to add reformatting of $sURL for local files; I suggested that this might be better as a separate function

> did not Return because @error is not set regadless of the fact that in browser I see: ERR_FILE_NOT_FOUND

And that's exactly how it should work IMO. We may just have to "agree to disagree" on this.

**Comment by mlipok** [comment]
I will answer in next 8 hours (going with son to cinema)

**Comment by Danp2** [comment]
> It looks like WebDriver at least chromedriver.exe does not give any error in this case.
Thus IMHO it is WD UDF role to check if the file exist.

 I disagree (again). The function provides a single feature -- Navigation. It doesn't attempt to verify / validate standard URLs. Why should local file URLs be any different? This is the responsibility of the calling routine.

**Comment by mlipok** [comment]
> It doesn't attempt to verify / validate standard URLs. Why should local file URLs be any different? This is the responsibility of the calling routine.

I disagree (again) ;)
try this code:
```
	_WD_Navigate($sSession, 'wwww.somenfakeurltotestwebdriverautoitudf.com')
	If @error Then Return
```
The @error wil be set and program return from the function.
And you will get:
```
__WD_Post: URL=HTTP://127.0.0.1:9515/session/e5f6fb06e97357a647a3cc235c57f9c6/url; $sData={"url":"wwww.somenfakeurltotestwebdriverautoitudf.com"}
__WD_Post: StatusCode=400; ResponseText={"value":{"error":"invalid argument","message":"invalid argument\n  (Session info: chrome=97.0.4692....
__WD_Post ==> Invalid argument: {"value":{"error":"invalid argument","message":"invalid argument\n  (Session info: chrome=97.0.4692.71)","stacktrace":"Backtrace:\n\tOrdinal0 [0x0088FDC3+2555331]\n\tOrdinal0 [0x008277F1+2127857]\n\tOrdinal0 [0x00722CD0+1060048]\n\tOrdinal0 [0x00715763+1005411]\n\tOrdinal0 [0x007145DD+1000925]\n\tOrdinal0 [0x007146B6+1001142]\n\tOrdinal0 [0x007243FA+1065978]\n\tOrdinal0 [0x00777BAD+1407917]\n\tOrdinal0 [0x00767B0C+1342220]\n\tOrdinal0 [0x007775FA+1406458]\n\tOrdinal0 [0x00767976+1341814]\n\tOrdinal0 [0x007436B6+1193654]\n\tOrdinal0 [0x00744546+1197382]\n\tGetHandleVerifier [0x00A29622+1619522]\n\tGetHandleVerifier [0x00AD882C+2336844]\n\tGetHandleVerifier [0x009223E1+541697]\n\tGetHandleVerifier [0x00921443+537699]\n\tOrdinal0 [0x0082D18E+2150798]\n\tOrdinal0 [0x00831518+2168088]\n\tOrdinal0 [0x00831660+2168416]\n\tOrdinal0 [0x0083B330+2208560]\n\tBaseThreadInitThunk [0x759B6739+25]\n\tRtlGetFullPathName_UEx [0x776A8AFF+1215]\n\tRtlGetFullPathName_UEx [0x776A8ACD+1165]\n"}}
_WD_Navigate: {"value":{"error":"invalid argument","message":"invalid argument\n  (Session info: chrome=97.0.4692.71)","stacktrace":"Backtrace:\n\tOrdinal0 [0x0088FDC3+2555331]\n\tOrdinal0 [0x008277F1+2127857]\n\tOrdinal0 [0x00722CD0+1060048]\n\tOrdinal0 [0x00715763+1005411]\n\tOrdinal0 [0x007145DD+1000925]\n\tOrdinal0 [0x007146B6+1001142]\n\tOrdinal0 [0x007243FA+1065978]\n\tOrdinal0 [0x00777BAD+1407917]\n\tOrdinal0 [0x00767B0C+1342220]\n\tOrdinal0 [0x007775FA+1406458]\n\tOrdinal0 [0x00767976+1341814]\n\tOrdinal0 [0x007436B6+1193654]\n\tOrdinal0 [0x00744546+1197382]\n\tGetHandleVerifier [0x00A29622+1619522]\n\tGetHandleVerifier [0x00AD882C+2336844]\n\tGetHandleVerifier [0x009223E1+541697]\n\tGetHandleVerifier [0x00921443+537699]\n\tOrdinal0 [0x0082D18E+2150798]\n\tOrdinal0 [0x00831518+2168088]\n\tOrdinal0 [0x00831660+2168416]\n\tOrdinal0 [0x0083B330+2208560]\n\tBaseThreadInitThunk [0x759B6739+25]\n\tRtlGetFullPathName_UEx [0x776A8AFF+1215]\n\tRtlGetFullPathName_UEx [0x776A8ACD+1165]\n"}}
_WD_Navigate ==> Invalid argument: HTTP status = 400
+>23:02:33 AutoIt3.exe ended.rc:0
+>23:02:33 AutoIt3Wrapper Finished.
>Exit code: 0    Time: 18.06
```

**Comment by mlipok** [comment]
> This is the responsibility of the calling routine.

Not quite because as I said before, WebDriver says everything is OK while the automated browser instance says ERR_FILE_NOT_FOUND.

Is this correct behavior of any software - to provide incorrect information?

**Comment by mlipok** [comment]
> * You want to add reformatting of $sURL for local files; I suggested that this might be better as a separate function

Could you explain why ?

**Comment by Danp2** [comment]
> As I said WebDriver currently behaves in a different way in case remote URL and local files. It looks like WebDriver properly check for remote URL but do not bother about checking local file.

I believe that you are mistaken. The reason an error was generated in your specific case is because you didn't pass an a valid URL as far as the webdriver was concerned. Try this instead (I see that you are editing your responses as I type this, which adds to the confusion) --

```
	_WD_Navigate($sSession, 'http://abcdesomeinvalidurlf.com')
	If @error Then Return
```

> Is this correct behavior of any software - to provide incorrect information?

IMO, It isn't providing incorrect information. It is performing the requested action and then responding that it succeeded, which it technically did.

> Could you explain why ?

Will respond to this in a bit

**Comment by mlipok** [comment]
> (I see that you are editing your responses as I type this, which adds to the confusion)
Sorry.


> Try this instead
> 
> ```
> 	_WD_Navigate($sSession, 'http://abcdesomeinvalidurlf.com')
> 	If @error Then Return
> ```


I get:
```
__WD_Post: URL=HTTP://127.0.0.1:9515/session/946474db93d0627919b965a8603a470f/url; $sData={"url":"http://abcdesomeinvalidurlf.com"}
__WD_Post: StatusCode=500; ResponseText={"value":{"error":"unknown error","message":"unknown error: net::ERR_NAME_NOT_RESOLVED\n  (Session i...
__WD_Post ==> Webdriver Exception: {"value":{"error":"unknown error","message":"unknown error: net::ERR_NAME_NOT_RESOLVED\n  (Session info: chrome=97.0.4692.71)","stacktrace":"Backtrace:\n\tOrdinal0 [0x002EFDC3+2555331]\n\tOrdinal0 [0x002877F1+2127857]\n\tOrdinal0 [0x00182E08+1060360]\n\tOrdinal0 [0x0017F9B1+1046961]\n\tOrdinal0 [0x001759BD+1006013]\n\tOrdinal0 [0x001765D0+1009104]\n\tOrdinal0 [0x00175C1E+1006622]\n\tOrdinal0 [0x0017527D+1004157]\n\tOrdinal0 [0x001743FA+1000442]\n\tOrdinal0 [0x001746B6+1001142]\n\tOrdinal0 [0x001843FA+1065978]\n\tOrdinal0 [0x001D7BAD+1407917]\n\tOrdinal0 [0x001C7B0C+1342220]\n\tOrdinal0 [0x001D75FA+1406458]\n\tOrdinal0 [0x001C7976+1341814]\n\tOrdinal0 [0x001A36B6+1193654]\n\tOrdinal0 [0x001A4546+1197382]\n\tGetHandleVerifier [0x00489622+1619522]\n\tGetHandleVerifier [0x0053882C+2336844]\n\tGetHandleVerifier [0x003823E1+541697]\n\tGetHandleVerifier [0x00381443+537699]\n\tOrdinal0 [0x0028D18E+2150798]\n\tOrdinal0 [0x00291518+2168088]\n\tOrdinal0 [0x00291660+2168416]\n\tOrdinal0 [0x0029B330+2208560]\n\tBaseThreadInitThunk [0x759B6739+25]\n\tRtlGetFullPathName_UEx [0x776A8AFF+1215]\n\tRtlGetFullPathName_UEx [0x776A8ACD+1165]\n\t(No symbol) [0x00000000]\n"}}
_WD_Navigate: {"value":{"error":"unknown error","message":"unknown error: net::ERR_NAME_NOT_RESOLVED\n  (Session info: chrome=97.0.4692.71)","stacktrace":"Backtrace:\n\tOrdinal0 [0x002EFDC3+2555331]\n\tOrdinal0 [0x002877F1+2127857]\n\tOrdinal0 [0x00182E08+1060360]\n\tOrdinal0 [0x0017F9B1+1046961]\n\tOrdinal0 [0x001759BD+1006013]\n\tOrdinal0 [0x001765D0+1009104]\n\tOrdinal0 [0x00175C1E+1006622]\n\tOrdinal0 [0x0017527D+1004157]\n\tOrdinal0 [0x001743FA+1000442]\n\tOrdinal0 [0x001746B6+1001142]\n\tOrdinal0 [0x001843FA+1065978]\n\tOrdinal0 [0x001D7BAD+1407917]\n\tOrdinal0 [0x001C7B0C+1342220]\n\tOrdinal0 [0x001D75FA+1406458]\n\tOrdinal0 [0x001C7976+1341814]\n\tOrdinal0 [0x001A36B6+1193654]\n\tOrdinal0 [0x001A4546+1197382]\n\tGetHandleVerifier [0x00489622+1619522]\n\tGetHandleVerifier [0x0053882C+2336844]\n\tGetHandleVerifier [0x003823E1+541697]\n\tGetHandleVerifier [0x00381443+537699]\n\tOrdinal0 [0x0028D18E+2150798]\n\tOrdinal0 [0x00291518+2168088]\n\tOrdinal0 [0x00291660+2168416]\n\tOrdinal0 [0x0029B330+2208560]\n\tBaseThreadInitThunk [0x759B6739+25]\n\tRtlGetFullPathName_UEx [0x776A8AFF+1215]\n\tRtlGetFullPathName_UEx [0x776A8ACD+1165]\n\t(No symbol) [0x00000000]\n"}}
_WD_Navigate ==> Webdriver Exception: HTTP status = 500

```

It still returns @error in case in case navigation has failed.

And this is my main thought.
I mean if navigation to a webpage or local file fails for whatever reason, the function should return an error.
But currently the error is only returned when the navigation leads to a website, but not when the navigation leads to a local file, which is not good. Therefore, I added checking if the file exists.

**Comment by mlipok** [comment]
> IMO, It isn't providing incorrect information. It is performing the requested action and then responding that it succeeded, which it technically did.

Technically even this following example:

> Try this instead
> 
> ```
> 	_WD_Navigate($sSession, 'http://abcdesomeinvalidurlf.com')
> 	If @error Then Return
> ```

did what they should:
image

but you get error:
```
__WD_Post: URL=HTTP://127.0.0.1:9515/session/33d6ba2b95630e25f91d74fba529bbd4/url; $sData={"url":"http://abcdesomeinvalidurlf.com"}
__WD_Post: StatusCode=500; ResponseText={"value":{"error":"unknown error","message":"unknown error: net::ERR_NAME_NOT_RESOLVED\n  (Session i...
__WD_Post ==> Webdriver Exception: {"value":{"error":"unknown error","message":"unknown error: net::ERR_NAME_NOT_RESOLVED\n  (Session info: chrome=97.0.4692.71)","stacktrace":"Backtrace:\n\tOrdinal0 [0x002EFDC3+2555331]\n\tOrdinal0 [0x002877F1+2127857]\n\tOrdinal0 [0x00182E08+1060360]\n\tOrdinal0 [0x0017F9B1+1046961]\n\tOrdinal0 [0x001759BD+1006013]\n\tOrdinal0 [0x001765D0+1009104]\n\tOrdinal0 [0x00175C1E+1006622]\n\tOrdinal0 [0x0017527D+1004157]\n\tOrdinal0 [0x001743FA+1000442]\n\tOrdinal0 [0x001746B6+1001142]\n\tOrdinal0 [0x001843FA+1065978]\n\tOrdinal0 [0x001D7BAD+1407917]\n\tOrdinal0 [0x001C7B0C+1342220]\n\tOrdinal0 [0x001D75FA+1406458]\n\tOrdinal0 [0x001C7976+1341814]\n\tOrdinal0 [0x001A36B6+1193654]\n\tOrdinal0 [0x001A4546+1197382]\n\tGetHandleVerifier [0x00489622+1619522]\n\tGetHandleVerifier [0x0053882C+2336844]\n\tGetHandleVerifier [0x003823E1+541697]\n\tGetHandleVerifier [0x00381443+537699]\n\tOrdinal0 [0x0028D18E+2150798]\n\tOrdinal0 [0x00291518+2168088]\n\tOrdinal0 [0x00291660+2168416]\n\tOrdinal0 [0x0029B330+2208560]\n\tBaseThreadInitThunk [0x759B6739+25]\n\tRtlGetFullPathName_UEx [0x776A8AFF+1215]\n\tRtlGetFullPathName_UEx [0x776A8ACD+1165]\n\t(No symbol) [0x00000000]\n"}}
_WD_Navigate: {"value":{"error":"unknown error","message":"unknown error: net::ERR_NAME_NOT_RESOLVED\n  (Session info: chrome=97.0.4692.71)","stacktrace":"Backtrace:\n\tOrdinal0 [0x002EFDC3+2555331]\n\tOrdinal0 [0x002877F1+2127857]\n\tOrdinal0 [0x00182E08+1060360]\n\tOrdinal0 [0x0017F9B1+1046961]\n\tOrdinal0 [0x001759BD+1006013]\n\tOrdinal0 [0x001765D0+1009104]\n\tOrdinal0 [0x00175C1E+1006622]\n\tOrdinal0 [0x0017527D+1004157]\n\tOrdinal0 [0x001743FA+1000442]\n\tOrdinal0 [0x001746B6+1001142]\n\tOrdinal0 [0x001843FA+1065978]\n\tOrdinal0 [0x001D7BAD+1407917]\n\tOrdinal0 [0x001C7B0C+1342220]\n\tOrdinal0 [0x001D75FA+1406458]\n\tOrdinal0 [0x001C7976+1341814]\n\tOrdinal0 [0x001A36B6+1193654]\n\tOrdinal0 [0x001A4546+1197382]\n\tGetHandleVerifier [0x00489622+1619522]\n\tGetHandleVerifier [0x0053882C+2336844]\n\tGetHandleVerifier [0x003823E1+541697]\n\tGetHandleVerifier [0x00381443+537699]\n\tOrdinal0 [0x0028D18E+2150798]\n\tOrdinal0 [0x00291518+2168088]\n\tOrdinal0 [0x00291660+2168416]\n\tOrdinal0 [0x0029B330+2208560]\n\tBaseThreadInitThunk [0x759B6739+25]\n\tRtlGetFullPathName_UEx [0x776A8AFF+1215]\n\tRtlGetFullPathName_UEx [0x776A8ACD+1165]\n\t(No symbol) [0x00000000]\n"}}
_WD_Navigate ==> Webdriver Exception: HTTP status = 500

```

So I say it again:

> WebDriver currently behaves in a different way in case remote URL and local files. It looks like WebDriver properly check for remote URL but do not bother about checking local file.
> 
> And I do not complain that it does not check, but that it does not give the correct result because it says everything is OK while the browser says ERR_FILE_NOT_FOUND.
> 
> So IMO this is a bug on the WebDriver side, and regardless of whether or when it gets fixed on the WebDriver side, we can prevent errors now by checking it on the WD UDF side.

here is why I mean in this case:

image

and result in console says:
```
__WD_Post: URL=HTTP://127.0.0.1:9515/session/e0660433d5f26027d99ed3a3e3d21ef0/url; $sData={"url":"file:///Z:/!!!_SVN_AU3/UDF_Forum/Other_Members/Test/WebDriver/NotExistenFile.html"}
__WD_Post: StatusCode=200; ResponseText={"value":null}...
_WD_Navigate: {"value":null}
+>00:05:18 AutoIt3.exe ended.rc:0
```

**Comment by Danp2** [comment]
When I tested this before, it didn't return an error when the URL was invalid. This may be due to the DNS provider in use at my current locale. I will have to perform some more tests when I get home.

**Comment by Danp2** [comment]
Tested with Firefox and found that it actually returns an error when the local file isn't found --
```
__WD_Post: StatusCode=200; ResponseText={"value":{"sessionId":"5d99a253-bb13-4c09-b25d-de156b26b577","capabilities":{"acceptInsecureCerts":t...
_WD_CreateSession: {"value":{"sessionId":"5d99a253-bb13-4c09-b25d-de156b26b577","capabilities":{"acceptInsecureCerts":true,"browserName":"firefox","browserVersion":"95.0.2","moz:accessibilityChecks":false,"moz:buildID":"20211218203254","moz:geckodriverVersion":"0.30.0","moz:headless":false,"moz:processID":30708,"moz:profile":"C:\\Users\\danpo\\AppData\\Local\\Temp\\rust_mozprofileC6tLVd","moz:shutdownTimeout":60000,"moz:useNonSpecCompliantPointerOrigin":false,"moz:webdriverClick":true,"pageLoadStrategy":"normal","platformName":"windows","platformVersion":"10.0","proxy":{},"setWindowRect":true,"strictFileInteractability":false,"timeouts":{"implicit":0,"pageLoad":300000,"script":30000},"unhandledPromptBehavior":"dismiss and notify"}}}
__WD_Post: URL=HTTP://127.0.0.1:4444/session/5d99a253-bb13-4c09-b25d-de156b26b577/url; $sData={"url":"file:///C:/Users/danpo/Dropbox/webdriver/help/Introduction2.htm"}
__WD_Post: StatusCode=500; ResponseText={"value":{"error":"unknown error","message":"Reached error page: about:neterror?e=fileNotFound&u=fil...
__WD_Post ==> Webdriver Exception: {"value":{"error":"unknown error","message":"Reached error page: about:neterror?e=fileNotFound&u=file%3A///C%3A/Users/danpo/Dropbox/webdriver/help/Introduction2.htm&c=UTF-8&d=Firefox%20can%E2%80%99t%20find%20the%20file%20at%20/C%3A/Users/danpo/Dropbox/webdriver/help/Introduction2.htm.","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:181:5\nUnknownError@chrome://remote/content/shared/webdriver/Errors.jsm:488:5\ncheckReadyState@chrome://remote/content/marionette/navigate.js:64:24\nonNavigation@chrome://remote/content/marionette/navigate.js:312:39\nemit@resource://gre/modules/EventEmitter.jsm:160:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteEventsParent.jsm:44:25\n"}}
_WD_Navigate: {"value":{"error":"unknown error","message":"Reached error page: about:neterror?e=fileNotFound&u=file%3A///C%3A/Users/danpo/Dropbox/webdriver/help/Introduction2.htm&c=UTF-8&d=Firefox%20can%E2%80%99t%20find%20the%20file%20at%20/C%3A/Users/danpo/Dropbox/webdriver/help/Introduction2.htm.","stacktrace":"WebDriverError@chrome://remote/content/shared/webdriver/Errors.jsm:181:5\nUnknownError@chrome://remote/content/shared/webdriver/Errors.jsm:488:5\ncheckReadyState@chrome://remote/content/marionette/navigate.js:64:24\nonNavigation@chrome://remote/content/marionette/navigate.js:312:39\nemit@resource://gre/modules/EventEmitter.jsm:160:20\nreceiveMessage@chrome://remote/content/marionette/actors/MarionetteEventsParent.jsm:44:25\n"}}
_WD_Navigate ==> Webdriver Exception: HTTP status = 500
```
So now we need to determine which behavior is correct.

**Comment by Danp2** [comment]
> > You want to add reformatting of $sURL for local files; I suggested that this might be better as a separate function
> 
> Could you explain why ?

Instead of blindly attempting to convert the URL without being asked, the script writer would need to explicitly call the function if they want to properly format a local file path. Using your method could lead to errors. For example, what if I passed `file:\\\c::\test.html` as the URL?

**Comment by mlipok** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=1285718

**Comment by mlipok** [comment]
Will you aproove 

```AutoIt
	If StringInStr($sURL, '\') Then
		$sURL = "file:///" & StringReplace($sURL, "\", "/")
	EndIf
```

instead
```AutoIt
	If StringInStr($sURL, '\') Then
		If Not FileExists($sURL) Then
			Return SetError(__WD_Error($sFuncName, $_WD_ERROR_NotFound, "File not exist: " & $sURL), 0, 0)
		EndIf
		$sURL = "file:///" & StringReplace($sURL, "\", "/")
	EndIf
```

If so I will commit but aslos create separate Issue for :


> I just checked with FireFox and I confirm that I get: `__WD_Post: StatusCode=500; ResponseText={"value":{"error":"unknown error","message":"Reached error page: about:neterror?e=fileNotFound&`
> 
> So this is ChromeDriver issue that he not return error in case of `ERR_FILE_NOT_FOUND`
> 
> Take a look here: https://bugs.chromium.org/p/chromium/issues/detail?id=1285718
> 
> I think we need to wait (here with this PR) for a response to my bug report.
> 
> Also here: [w3c/webdriver#1638](https://github.com/w3c/webdriver/issues/1638)

**Comment by Danp2** [comment]
I propose that we close this PR without merging it and move the discussion to a new issue.

**Comment by mlipok** [comment]
I was in hope that after [change of approach](https://github.com/Danp2/WebDriver/pull/124/commits/47d84dc881a505442ba0b85d5cea734e6eed338c) 
You will approve this PR.

**Comment by Danp2** [comment]
> I was in hope that after change of approach You will approve this PR.

> I hope to make you follow the same approach in this UDF as well.

I think it's hopeless. 😁 

Seriously, we have gone round and round on this, and I'm still not convinced that _WD_Navigate should be responsible for modifying the URL being passed to the webdriver without the request or permission of the script writer.

**Comment by mlipok** [comment]
Ok.

**Pull request closed without merging**